src/structlog/processors.py

```diff
@@ -909,3 +911,75 @@ def __call__(
                 event_dict["event"] = replace_by
 
         return event_dict
+
+
+class CallsiteNamespaceAddr:
+    """
+    Attempt to identify and add the caller namespace to the event dict
+    under the ``namespace`` key.
+
+    Arguments:
+
+        levels:
+            A set of log levels to add the ``namespace`` key and
+            information to. An empty set == *
+
+    .. versionadded:: 23.3.0
+    """
+
+    def __init__(self, levels: set[str] | list[str] | None = None):
+        self.levels = levels
+
+    def __call__(
+        self, logger: WrappedLogger, name: str, event_dict: EventDict
+    ) -> EventDict:
+        if self.levels and name not in self.levels:
+            return event_dict
+
+        f, _ = _find_first_app_frame_and_name()
+        event_dict["namespace"] = self.get_qual_name(f)
+
+        return event_dict
+
+    def get_qual_name(self, frame: FrameType) -> str:
+        """
+            For a given app frame, attempt to deduce the namespace
+            by crawling through the frame's ``f_globals`` to find
+            matching object code.
+
+            This O(n) procedure should return as O(1) in most situations,
+            but buyer beware.
+
+            Arguments:
+
+                frame:
+                    Frame to process.
+
+            Returns:
+
+                string of the deduced namespace
+
+        .. versionadded:: 23.3.0
+        """
+        identified_namespace = frame.f_code.co_name
+
+        # pull out classes from the frames `f_globals` for testing against
+        cls_queue = deque(
+            obj for obj in frame.f_globals.values() if inspect.isclass(obj)
+        )
+        namespace_found = False
+
+        while cls_queue and not namespace_found:
+            cls = cls_queue.popleft()
+            member = getattr(cls, frame.f_code.co_name, None)
+            # store the current namespace as a fall back (probably the namespace)
+            identified_namespace = f"{cls.__module__}.{frame.f_code.co_name}"
+            if inspect.isfunction(member) and member.__code__ == frame.f_code:
+                identified_namespace = (
+                    f"{member.__module__}.{member.__qualname__}"
+                )
+                # we found our code match, can stop looking
+                namespace_found = True
```

Comment:
A simple `break` should do, no?

Reply:
I reverted back to `break` based logic

Reply:
```
Name                       Stmts   Miss Branch BrPart  Cover   Missing
----------------------------------------------------------------------
src/structlog/_frames.py      47      0     15      1    98%   104->115
----------------------------------------------------------------------
TOTAL                       1781      0    462      1    99%

18 files skipped due to complete coverage.
Coverage failure: total of 99 is less than fail-under=100
Error: Process completed with exit code 2.
```
There is the coverage fault on the `break` logic.

Reply:
I got creative and worked around it without needing to `nocover` it.

Reply:
What this is telling you means that you're missing _branch_ coverage. I suspect it means that the loop never loops twice (aka: the break is always executed)?

Reply:
Correct — which is why the “should operate at O(1)” comment… loop is to be thorough for edge cases.

I got it cleaned and passing

Reply:
OK, so this is actually great, because I honestly don't quite understand what/how the code is doing.

Could you please in plain words explain to me what it's doing? Finding a test case where it needs more than one iteration is very helpful for my understanding.